### PR TITLE
Migrate Photon to Typescript

### PR DESCRIPTION
**Pull request by noahtallen**
### Changes proposed in this Pull Request

Inspired by #59031, where @sirbrillig ports a JS package to Typescript. We're doing this because these packages do not have any type definitions, which means they cause errors when imported into TS files in the Calypso client. Now, we could just add typedefs -- but where's the fun in that!? :)

### Testing instructions
1. Load calypso.live
2. Visit `/themes/$site` for a simple site
3. Verify that the screenshots of themes in the list still load.

Each image in the list goes through photon:
image

### CODE REVIEW

File: packages/photon/src/index.ts [hunk 1]
```diff
@@ -7,17 +7,36 @@ const debug = debugFactory( 'photon' );
 /**
  * Options argument to query string parameter mappings.
  */
-const mappings = {
-	width: 'w',
-	height: 'h',
-	letterboxing: 'lb',
-	removeLetterboxing: 'ulb',
-};
+function optionToQueryParam( key: string ) {
```
Comment (noahtallen) [inline packages/photon/src/index.ts]: Typescript didn't like indexing the mapping object, so the switch statement makes the types a bit happier.

File: packages/photon/src/index.ts [hunk 2]
```diff
@@ -7,17 +7,36 @@ const debug = debugFactory( 'photon' );
...
 
+type PhotonOpts = {
+	width?: number;
+	height?: number;
+	hostname?: string;
+	host?: string;
+	secure?: boolean;
+	zoom?: number;
+	resize?: string;
+	fit?: string;
```
Comment (noahtallen) [inline packages/photon/src/index.ts:29]: I copied options in as they seem to be used in the rest of the codebase.

File: packages/photon/src/index.ts [hunk 3]
```diff
@@ -64,25 +83,21 @@ export default function photon( imageUrl, opts ) {
 		photonUrl.pathname = formattedUrl;
 		photonUrl.hostname = serverFromUrlParts( formattedUrl, photonUrl.protocol === 'https:' );
 		if ( wasSecure ) {
-			photonUrl.searchParams.set( 'ssl', 1 );
+			photonUrl.searchParams.set( 'ssl', '1' );
```
Comment (noahtallen) [inline packages/photon/src/index.ts:80]: `searchParams` only supports string values. I think it _technically_ converts to strings under the hood, but since it's not officially specified that way, Typescript only allows string values here.

File: packages/photon/src/index.ts [hunk 4]
```diff
@@ -64,25 +83,21 @@ export default function photon( imageUrl, opts ) {
...
 		}
 	}
 
 	if ( opts ) {
-		for ( const i in opts ) {
-			// allow configurable "hostname"
-			if ( i === 'host' || i === 'hostname' ) {
-				photonUrl.hostname = opts[ i ];
+		for ( const [ opt, value ] of Object.entries( opts ) ) {
```
Comment (noahtallen) [inline packages/photon/src/index.ts:85]: Iterating over the entries means we can avoid casting `i` to `keyof typeof opts`.

File: packages/photon/src/index.ts [hunk 5]
```diff
@@ -64,25 +83,21 @@ export default function photon( imageUrl, opts ) {
...
-
-			// allow non-secure access
-			if ( i === 'secure' && ! opts[ i ] ) {
+			if ( opt === 'secure' && ! value ) {
 				photonUrl.protocol = 'http:';
 				continue;
 			}
-
-			photonUrl.searchParams.set( mappings[ i ] || i, opts[ i ] );
+			photonUrl.searchParams.set( optionToQueryParam( opt ), value.toString() );
```
Comment (noahtallen) [inline packages/photon/src/index.ts]: Again, `searchParams` requires string input.

File: packages/photon/src/decs.d.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+declare module 'crc32' {
+	export default function crc32( s: string ): string;
```
Comment (noahtallen) [inline packages/photon/src/decs.d.ts:2]: Unfortunately no type package for this one -- looks like it hasn't been updated in 10 years. :eyes:

File: packages/photon/package.json [hunk 7]
```diff
@@ -37,11 +42,12 @@
...
 	"devDependencies": {
-		"@automattic/calypso-typescript-config": "workspace:^"
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"@types/seed-random": "^2.2.1"
 	},
 	"scripts": {
-		"clean": "rm -rf dist",
-		"build": "transpile",
+		"clean": "tsc --build ./tsconfig.json ./tsconfig-cjs.json --clean && rm -rf dist",
+		"build": "tsc --build ./tsconfig.json ./tsconfig-cjs.json && copy-assets",
```
Comment (scinos) [inline packages/photon/package.json]: `copy-assets` is only needed if the package has style sheets or images, which is not the case.

File: packages/photon/package.json [hunk 8]
```diff
@@ -37,11 +42,12 @@
 		"seed-random": "^2.2.0"
 	},
 	"devDependencies": {
-		"@automattic/calypso-typescript-config": "workspace:^"
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"@types/seed-random": "^2.2.1"
```
Comment (scinos) [inline packages/photon/package.json]: `@types/foo` should be in the same group as `foo` (https://github.com/Automattic/wp-calypso/pull/58711#discussion_r760901539)

Comment (noahtallen) [reply packages/photon/package.json]: That's very useful to know, thanks!

File: packages/photon/package.json [hunk 9]
```diff
@@ -40,8 +46,8 @@
 		"@automattic/calypso-typescript-config": "workspace:^"
```
Comment (scinos) [inline packages/photon/package.json]: Missing dependency on `typescript`?

Comment (noahtallen) [reply packages/photon/package.json]: yep! Relatedly, looks like we missed it for the `viewport` migration as well.

Comment (noahtallen) [reply packages/photon/package.json]: Should `@automattic/calypso-typescript-config` have a peer dependency on typescript?

File: apps/editing-toolkit/package.json [hunk 10]
```diff
@@ -60,6 +60,7 @@
 		"@automattic/plans-grid": "workspace:^",
 		"@automattic/tour-kit": "workspace:^",
 		"@automattic/typography": "workspace:^",
+		"@automattic/viewport": "workspace:^",
```
Comment (noahtallen) [inline apps/editing-toolkit/package.json:63]: Drive-by fix for a missing dep in ETK

File: packages/viewport/package.json [hunk 11]
```diff
@@ -31,6 +31,7 @@
 		"access": "public"
 	},
 	"devDependencies": {
-		"@automattic/calypso-typescript-config": "workspace:^"
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"typescript": "^4.5.3"
```
Comment (noahtallen) [inline packages/viewport/package.json:35]: drive-by fix for missing TS dep in viewport

File: packages/photon/package.json (hunk 9 above)
Comment (scinos) [reply packages/photon/package.json]: `@automattic/calypso-typescript-config` exports a JS-only config as well. The name is misleading :(

File: packages/photon/src/index.ts (hunk 1 above)
Comment (jsnajdr) [reply packages/photon/src/index.ts]: Typing the `mappings` object as `Record<string, string>` works for me. The expression `mappings[key] ?? key` is then type-safe.

File: packages/photon/src/index.ts (hunk 2 above)
Comment (jsnajdr) [reply packages/photon/src/index.ts:29]: There are also the `letterboxing` and `removeLetterBoxing` options -- they have a query param mapping.

File: packages/photon/src/index.ts [hunk 12]
```diff
@@ -26,11 +45,11 @@ const PHOTON_BASE_URL = 'https://i0.wp.com';
...
  *
- * @param {string} imageUrl - the URL of the image to run through Photon
- * @param {object} [opts] - optional options object with Photon options
- * @returns {string} The generated Photon URL string
+ * @param imageUrl - the URL of the image to run through Photon
+ * @param [opts] - optional options object with Photon options
+ * @returns The generated Photon URL string
  */
-export default function photon( imageUrl, opts ) {
+export default function photon( imageUrl: string, opts?: PhotonOpts ): string | null {
```
Comment (jsnajdr) [inline packages/photon/src/index.ts:46]: We could change the return type to just `string` and if the `imageUrl` fails to parse, just return it as it is.

All `photon()` usages in the codebase assume that the function returns an URL string and never check for success.

Comment (noahtallen) [reply packages/photon/src/index.ts:46]: That makes sense to me. I wasn't sure if it was used outside of Calypso, in which case this would be a breaking change.

Comment (noahtallen) [reply packages/photon/src/index.ts:46]: Actually, it looks like [safeImageUrl()](https://github.com/Automattic/wp-calypso/blob/31d3844d216c261ed9bff5ea0abadc0aa0d78b4b/client/lib/safe-image-url/index.js#L45) uses `photon()` under the hood. Other places in the codebase rely on `safeImageUrl` returning null if the image won't work: https://github.com/Automattic/wp-calypso/blob/31d3844d216c261ed9bff5ea0abadc0aa0d78b4b/client/lib/post-normalizer/rule-keep-valid-images.js#L6

Comment (jsnajdr) [reply packages/photon/src/index.ts:46]: OK, let's keep the `string | null` return type then.

**Comment by noahtallen** [comment]
I had to revert this because Typescript was updated on trunk, but not in this branch. So the lockfile was modified after merging, causing all the build tasks to fail. I should have rebased it again today before merging :)
